### ISSUE TITLE
Clamd (Clam AV Daemon) fails to detect Eicar file with the string in Eicar file having trailing whitespaces / tabs as Affected

### ISSUE THREAD

**Issue by Meet11** [Cisco-Talos/clamav#1097]
Clamd (Clam AV Daemon) fails to detect the Eicar file with the string in the Eicar file having `trailing spaces` as Affected but passing the same file as input to Clamscan.exe marks the file as Affected.

**Steps to reproduce -**
1. Create a .txt file with the content provided below
`X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*`
Append "space" at the end of the content

2. clamdscan <name_of_file.txt> - The file is marked 'OK'

3. clamscan.exe <name_of_file.txt> - The file is marked 'FOUND'

Eicar.org's official documentation mentions that the Eicar file's string having trailing whitespaces is still a valid Eicar file and should be detected by anti virus - [https://www.eicar.org/download-anti-malware-testfile/#:~:text=X5O!P%25%40AP,the%20digit%20zero](https://www.eicar.org/download-anti-malware-testfile/#:%7E:text=X5O!P%25%40AP,the%20digit%20zero)

**Comment by rsundriyal** [comment]
@Meet11 We are not seeing the behavior in the latest ClamAV version 1.2.1 (In both windows/Linux).

 Can you share some more information about,

- [ ] Version of ClamAV (both clamscan/clamd if different)
- [ ] databases used by both clamd and clamscan, if different?
- [ ] Also, I tried appending spaces and tabs in the EICAR file but was not able to reproduce it. Can you share your test input file as well?

**Comment by Meet11** [comment]
@rsundriyal 

Version of Clam AV - 0.103.9
Same Database being used.
[eicar_test.txt](https://github.com/Cisco-Talos/clamav/files/13418841/eicar_test.txt)

An update to my observations -

For the above-mentioned results, I was running clamd process on a container where I am observing the issue. It is an ubuntu 22.04 container.

I have further tried executing with the same inputs in a different environment (Windows as well as on Linux) with Clam av 0.103.9 and am observing expected results where the Eicar file having trailing spaces is getting marked as affected.

The issue is specific to the container itself, though, I have kept clamd.conf same across the environments (for Linux)
I am now trying to identify what might be causing the issue within the container.

Note: Both clamscan.exe and clamdscan are marking the Eicar file as not affected in the container.

Will update.

**Comment by micahsnyder** [comment]
@Meet11 ClamAV's ability to detect allowed variations of EICAR relies on a bytecode signature from bytecode.cvd. 

If you run `clamconf -n` it should include a list of the available databases, like this:
```

Database information
--------------------
Database directory: /var/lib/clamav
daily.cld: version 27099, sigs: 2047099, built on Mon Nov 20 03:39:02 2023
[3rd Party] clamav.hdb: 1 sig
main.cvd: version 62, sigs: 6647427, built on Thu Sep 16 08:32:42 2021
bytecode.cvd: version 334, sigs: 91, built on Wed Feb 22 16:33:21 2023
Total number of signatures: 8694618
```

Can you confirm if bytecode.cvd is present?

**Comment by Meet11** [comment]
@micahsnyder 

That's right, the container was missing bytecode.cvd. Adding bytecode.cvd has resolved the issue.

**Comment by Meet11** [comment]
@micahsnyder 

Just an update and a question here -

bytecode.cvd was removed from the container env. and set to no in freshclam.conf because of the following issue and as the solution to it - 

https://forum.iredmail.org/topic1043-clamav-keeps-crashing.html

**Link to highlight -** 
https://forum.iredmail.org/topic1043-clamav-keeps-crashing.html#:~:text=I%20searched%20ClamAV%20bugzilla%20and%20this%20bug%20has%20already%20been%20detected.%20The%20easiest%20solution%20is%20to%20delete%20/var/lib/clamav/bytecode.cld%20and%20specify%20%22Bytecode%20off%22%20in%20/etc/clamav/freshclam.conf.%20This%20way%20the%20JIT%20has%20no%20definition%20and%20PAX%20doesn%27t%20kick%20in.

This bug was filed on [buzilla](https://wwws.clamav.net/bugzilla/show_bug.cgi?id=2092) but bugzilla server is no longer running.
Are you aware if this was fixed or know a place where I can look this up?

**Comment by micahsnyder** [comment]
@Meet11 sorry I didn't follow up afterwards.  

You will have to re-enable bytecode.cvd if you want full detection capabilities including detecting variations allowed on the EICAR file.

Our old bugzilla server was decommissioned, though we maintain a copy on our internal network for reference.  The bug was apparently "fixed" by updating the warning message printed on startup to tell you to run `paxctl -cm <executable>` so that PaX will allow RWX mapping. 

The warning message code [for PaX is here](https://github.com/Cisco-Talos/clamav/blob/0d4c99465e005ded3926b9e8631518f441fa6aa1/libclamav/builtin_bytecodes.h#L99-L100C65) 

If you're facing the same problem [with SELinux, this message](https://github.com/Cisco-Talos/clamav/blob/0d4c99465e005ded3926b9e8631518f441fa6aa1/libclamav/builtin_bytecodes.h#L95-L96) says to run `setsebool -P clamd_use_jit on` to allow JIT. 

This code in clam looks like it is commented out, but it is just a copy of [the same bytecode code that is hardcoded into clam](https://github.com/Cisco-Talos/clamav/blob/0d4c99465e005ded3926b9e8631518f441fa6aa1/libclamav/builtin_bytecodes.h#L32-L63).  

If SELinux or PaX is blocking the JIT compiler used when loading bytecode signatures, it should have printed one of these messages when you tried starting clamd with `bytecode.cvd`.

**Comment by micahsnyder** [comment]
Come to think of it, this rings a bell from when I first started working on ClamAV. You may also need:
`setsebool -P antivirus_can_scan_system 1`

Ref: https://bugzilla.redhat.com/show_bug.cgi?id=1449616#c4

**Comment by micahsnyder** [comment]
Also ref: https://github.com/Cisco-Talos/clamav/pull/89

It seems like I misunderstood the change suggested in https://github.com/Cisco-Talos/clamav/pull/89, and that it was just to provide a warning message to the user with instructions on what to do if SELinux is blocking the JIT compiler.  🤦

**Comment by dgm** [comment]
at the moment, none of my clamav installs are detecting the eicar signature.

 # cat eicar.txt
X5O!P%@AP[4\PZX54(P^)7CC)7}-STANDARD-ANTIVIRUS-TEST-FILE!+H*
/ # clamdscan eicar.txt
//eicar.txt: OK

----------- SCAN SUMMARY -----------
Infected files: 0
Time: 0.007 sec (0 m 0 s)
Start Date: 2025:02:17 18:50:53
End Date:   2025:02:17 18:50:53



Database directory: /var/lib/clamav
bytecode.cvd: version 335, sigs: 86, built on Tue Feb 27 15:37:24 2024
daily.cvd: version 27552, sigs: 2072975, built on Mon Feb 17 09:47:21 2025
main.cvd: version 62, sigs: 6647427, built on Thu Sep 16 12:32:42 2021
Total number of signatures: 8720488

**Comment by val-ms** [comment]
@dgm the eicar test file text is:
`X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*`
not:
`X5O!P%@AP[4\PZX54(P^)7CC)7}-STANDARD-ANTIVIRUS-TEST-FILE!+H*`
(second one is from your log)

It seems like your test file was mangled by a shell that evaluated `$EICAR` as an empty string, and $H as an empty string. 

You can also get the file from https://www.eicar.org/download-anti-malware-testfile/

**Comment by val-ms** [comment]
Closing this issue because the original reasons for the ticket were resolved as far as I can tell, and no response on the resolution from when it was reopened.

**Comment by millerick** [comment]
I ran into this as well.  One of our environments stopped reporting the EICAR file as a virus when there was a trailing newline, but did report it when I removed the trailing newline.

**Comment by val-ms** [comment]
@millerick I have no idea. That works okay for me. 
```
❯ echo 'X5O!P%@AP[4\PZX54(P^)7CC)7}$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!$H+H*' > eicar.txt

❯ ~/clams/1.4.1/bin/clamdscan ~/eicar.txt
/home/val/eicar.txt: Eicar-Signature FOUND

----------- SCAN SUMMARY -----------
Infected files: 1
Time: 0.003 sec (0 m 0 s)
Start Date: 2025:03:25 15:59:56
End Date:   2025:03:25 15:59:56
```

**Comment by millerick** [comment]
It still works for me on some of our clusters, but not all of them.  I do think there is some type of inconsistency in `clamav`'s behavior.

**Comment by balex25** [comment]
If someone has this problem, it seems there is a solution. Edit `sudo nano /etc/clamd.d/scan.conf` and change from `User clamscan` to `User apache` or (or `User nginx`, depending on your setup).

This fixed my problem, and it took me about one day to figure out why it wasn't detecting the EICAR text virus from the PHP scan. This seems to be a problem with clamscan user permissions...
